metadata/src/main/java/org/apache/kafka/image/ProducerIdsImage.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.image;
+
+import org.apache.kafka.common.metadata.ProducerIdsRecord;
+import org.apache.kafka.server.common.ApiMessageAndVersion;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Consumer;
+
+
+/**
+ * Stores the highest seen producer ID in the metadata image.
+ *
+ * This class is thread-safe.
+ */
+public final class ProducerIdsImage {
+    public final static ProducerIdsImage EMPTY = new ProducerIdsImage(-1L);
+
+    private final long highestSeenProducerId;
+
+    public ProducerIdsImage(long highestSeenProducerId) {
+        this.highestSeenProducerId = highestSeenProducerId;
+    }
+
+    public long highestSeenProducerId() {
+        return highestSeenProducerId;
+    }
+
+    public void write(Consumer<List<ApiMessageAndVersion>> out) {
+        if (highestSeenProducerId >= 0) {
+            out.accept(Collections.singletonList(new ApiMessageAndVersion(
+                new ProducerIdsRecord().
+                    setBrokerId(-1).
+                    setBrokerEpoch(-1).
+                    setProducerIdsEnd(highestSeenProducerId), (short) 0)));
```

Comment:
Okay. Is it fair to assume that we will revisit this code regarding the version in `ApiMessageAndVersion`as part of KIP-778?

Reply:
yes, we will need to have some versioning code in after KIP-778, so that we generate a snapshot at the appropriate `metadata.version`. Probably this will take the form of adding a `version` parameter to the snapshot generation function. cc @mumrah